batch/batch/driver/instance_collection/pool.py

```diff
@@ -566,10 +626,16 @@ async def user_runnable_jobs(user, remaining):
             ):
                 async for record in self.db.select_and_fetchall(
                     '''
-SELECT job_id, spec, cores_mcpu
-FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_inst_coll_cancelled)
-WHERE batch_id = %s AND state = 'Ready' AND always_run = 1 AND inst_coll = %s
-LIMIT %s;
+SELECT t.* FROM (
+  SELECT jobs.job_id, spec, cores_mcpu, JSON_ARRAYAGG(region) AS regions, COUNT(region) AS n_regions
+  FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_inst_coll_cancelled)
+  LEFT JOIN job_regions ON jobs.batch_id = job_regions.batch_id AND jobs.job_id = job_regions.job_id
+  LEFT JOIN region_ids ON job_regions.region_id = region_ids.region_id
+  WHERE jobs.batch_id = %s AND state = 'Ready' AND always_run = 1 AND inst_coll = %s
+  GROUP BY jobs.job_id, spec, cores_mcpu
+  ORDER BY jobs.job_id
+  LIMIT %s ) AS t
+ORDER BY -n_regions DESC, JSON_UNQUOTE(regions), job_id;
```

Comment:
How does this order by treat jobs with no specified regions? Those should go at the end

Reply:
So there's no coalesce on the count to compute the n_regions. This way no specified regions should have n_regions be NULL. With `-n_regions DESC`, we're sorting in ascending order, but nulls are last.